### ISSUE TITLE
Diagnostic source markers in the editor gutter get muddled when both "Run diagnostics in background" and "Check code for problems when saving" are checked

### ISSUE THREAD

**Issue by jgutman** [rstudio/rstudio#10430]
### System details

    RStudio Edition : Desktop
    RStudio Version : 2022.06.0-daily+26
    OS Version      :  macOS Monterey 12.1, aarch64-apple-darwin20 (64-bit)
    R Version       :  R-devel (4.2 2022-01-11 r81471)

### Steps to reproduce the problem

The behavior is wonky and difficult to pin down, but here's my best attempt and reproducing the behavior. 

1. Create a file with a bunch of diagnostic issues:

```
add_numbers <- function(x,y) {
  x + y
}

add_numbers(1)

list(
  first = 1
  second = 2
)

hw <- HoltWinters(ldeaths)
p <- predict(HW, n.ahead = 36, level = 0.95)

add_two <- function(number_1, number_2) {
  result <- number_1 + number_2
  return(number_1 + number_2)
}

x<-2+1   # comment
```

2. In Tools > Global Options > Code Diagnostics, select the following options
![image](https://user-images.githubusercontent.com/7817881/151186452-bfbf48e4-64bb-4df4-ac77-dbf00f651ef0.png)

Diagnostic markers appear in the gutter, on lines 5, 8, 13, 16, and 20. Great.

3. Edit the file, adding 2 blank lines after line 11. Save the file. Restart R Session (Shift + Cmd + 0).

Now the source markers in the gutter are all messed up. All markers now appear twice. Old line 5 and old line 8 are still on lines 5 and 8, and the markers on those lines are duplicated (one with unrendered symbols, one correctly). Old line 13 and old line 20 are still there, despite the associated code having been moved, and the symbols are not rendered correctly, but they also appear on lines 15 and 22, where they should be, and rendered correctly.

![image](https://user-images.githubusercontent.com/7817881/151188258-e8d3c073-6fdc-45eb-b68d-79d03a6875a8.png)
![image](https://user-images.githubusercontent.com/7817881/151188283-20e6d30d-db2d-467f-8e1e-e8db38af35c8.png)

I was also able to produce the same issue even with "Run diagnostics in background" unchecked and "Check code for problems when saving", with or without refreshing the R session. 

### Describe the problem in detail

There's a weird persistence and duplication of source markers when refreshing the R session. Sometimes, it seems like this can happen even without refreshing the R session. Sometimes, it seems like this can happen even if only one of "Check code for problems when saving" or "Run diagnostics in background" are checked.

### Describe the behavior you expected

Source markers should render correctly and only display on the line of code where the associated code now lives.

- [X] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [X] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [X] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by jgutman** [comment]
cc @kevinushey who might be familiar with what's going on here

**Comment by jgutman** [comment]
This behavior does not exist in 2022.02.0-preview+405 or 2022.06.0-daily+13. It appears to be introduced in 2022.06.0-daily+14

**Comment by astayleraz** [comment]
This looks like it came up in our automation https://build.rstudioservices.com/job/IDE/job/qa-automation/1831/testReport/rstudio_server_pro.tests.test_VisualMarkdownEditor/TestVisualMarkdownEditor/test_linting_error_appears_in_gutter/
We are expecting what shows on the left which is  Prairie Trillium and now get what is on the right with Spotten Wakerobin. 
![Screen Shot 2022-01-26 at 10 39 59 AM](https://user-images.githubusercontent.com/1482677/151220117-783d03f7-1ce6-445d-9a82-f8dde6e693c5.png)

**Comment by timtmok** [comment]
I've reproduced it successfully now. Restarting R definitely triggers it. The code diagnostics has to have run so it produces source markers. Restarting R likely triggers the linting code and adds the source markers for the gutter, which creates the duplicates.

**Comment by jgutman** [comment]
Note that we can see this duplicate issue in the diagnostics under the visual markdown editor as well. 

![image](https://user-images.githubusercontent.com/7817881/152039244-20c9d980-abc3-4e30-b2ff-c18298e28069.png)

Code:
~~~
{r}
print(b)
b <- "hello world!"
~~~

when editing the markdown in Visual mode and restarting the R session

**Comment by jgutman** [comment]
Verified this is fixed in RStudio Desktop 2022.06.0+76, tested on Mac OS 12.2 with R-devel 4.2.0
